Feat: statsd (#874)
Kong/developer.konghq.com: app/_kong_plugins/statsd/index.md

  - collectd
---

> Log [metrics](#metrics) for a Service or Route to a StatsD server.

Reviewer: ```suggestion
The StatsD plugin logs [metrics](#metrics) for a [Gateway Service](/gateway/entities/service/) or [Route](/gateway/entities/route/) to a StatsD server.
```

> It can also be used to log metrics on [Collectd](https://collectd.org/)

Reviewer: ```suggestion
It can also be used to log metrics on the [Collectd](https://collectd.org/)
```

By default, the plugin sends a packet for each metric it observes. The `udp_packet_size` option
configures the greatest datagram size the plugin can combine. It should be less than
65507 according to UDP protocol. Consider the MTU of the network when setting this parameter.

## Queueing

The StatsD plugin uses a queue to decouple the production and
consumption of data. This reduces the number of concurrent requests
made to the upstream server under high load situations and provides
buffering during temporary network or upstream outages.

You can set several parameters to configure the behavior and capacity
of the queues used by the plugin. For more information about how to
use these parameters, see
[Plugin Queuing Reference](/gateway/latest/kong-plugins/queue/reference/)
in the {{site.base_gateway}} documentation.

The queue parameters all reside in a record under the key `queue` in
the `config` parameter section of the plugin.

Queues are not shared between workers and queueing parameters are
scoped to one worker.  For whole-system capacity planning, the number
of workers need to be considered when setting queue parameters.

> ## Metrics

Reviewer: ```suggestion
## Metrics

The following configure the metrics that will be logged:
```

<!-- vale off -->
{% table %}
columns:
  - title: Metric
    key: metric
  - title: Description
    key: description
  - title: Namespace syntax
    key: namespace
rows:
  - metric: request_count
    description: The number of requests.
    namespace: "`kong.service.<service_identifier>.request.count`"
  - metric: request_size
    description: The request's body size in bytes.
    namespace: "`kong.service.<service_identifier>.request.size`"
  - metric: response_size
    description: The response's body size in bytes.
    namespace: "`kong.service.<service_identifier>.response.size`"
  - metric: latency
    description: The time interval in milliseconds between the request and response.
    namespace: "`kong.service.<service_identifier>.latency`"
  - metric: status_count
    description: Tracks each status code returned in a response.
    namespace: "`kong.service.<service_identifier>.status.<status>`"
  - metric: unique_users
    description: Tracks unique users who made requests to the underlying Service or Route.
    namespace: "`kong.service.<service_identifier>.user.uniques`"
  - metric: request_per_user

>     description: Tracks the request count per consumer.

Reviewer: ```suggestion
    description: Tracks the request count per Consumer.
```

    namespace: "`kong.service.<service_identifier>.user.<consumer_identifier>.request.count`"
  - metric: upstream_latency
    description: Tracks the time in milliseconds it took for the final Service to process the request.
    namespace: "`kong.service.<service_identifier>.upstream_latency`"
  - metric: kong_latency

>     description: Tracks the internal Kong latency in milliseconds that it took to run all the plugins.

Reviewer: ```suggestion
    description: Tracks the internal {{site.base_gateway}} latency in milliseconds that it took to run all the plugins.
```

    namespace: "`kong.service.<service_identifier>.kong_latency`"
  - metric: status_count_per_user

>     description: Tracks the status code per consumer per Service.

Reviewer: ```suggestion
    description: Tracks the status code per Consumer per Service.
```

    namespace: "`kong.service.<service_identifier>.user.<consumer_identifier>.status.<status>`"
  - metric: status_count_per_workspace

>     description: The status code per workspace.

Reviewer: ```suggestion
    description: The status code per Workspace.
```

  - metric: shdict_usage
    description: >-
      The usage of a shared dict, sent once every minute.<br><br>
      Monitors any `lua_shared_dict` used by {{site.base_gateway}}. You can find all the shared dicts {{site.base_gateway}} has configured using the [`/status`](/gateway/api/admin-ee/latest/#/operations/get-status) endpoint of the Admin API.<br><br>
      For example, the metric might report on `shdict.kong_locks` or `shdict.kong_counters`.
    namespace: "`kong.node.<node_hostname>.shdict.<lua_shared_dict>.free_space`<br><br>`kong.node.<node_hostname>.shdict.<lua_shared_dict>.capacity`"
  - metric: cache_datastore_hits_total
    description: The total number of cache hits. ({{site.ee_product_name}} only)
    namespace: "`kong.service.<service_identifier>.cache_datastore_hits_total`"
  - metric: cache_datastore_misses_total
    description: The total number of cache misses. ({{site.ee_product_name}} only)
    namespace: "`kong.service.<service_identifier>.cache_datastore_misses_total`"
{% endtable %}

<!-- vale on -->
If a request URI doesn't match any Routes, the following metrics are sent instead:
<!-- vale off -->

> Metric                     | Description | Namespace
> ---                        | ---         | ---
> `request_count`            | The request count. | `kong.global.unmatched.request.count`
> `request_size`             | The request's body size in bytes. | `kong.global.unmatched.request.size`
> `response_size`            | The response's body size in bytes. | `kong.global.unmatched.response.size`
> `latency`                  | The time interval between when the request started and when the response is received from the upstream server. | `kong.global.unmatched.latency`
> `status_count`             | The status count. | `kong.global.unmatched.status.<status>.count`
> `kong_latency`             | The internal Kong latency in milliseconds that it took to run all the plugins. | `kong.global.unmatched.kong_latency`

Reviewer: ```suggestion
| Metric                     | Description | Namespace |
| ---                        | ---         | --- |
| `request_count`            | The request count. | `kong.global.unmatched.request.count` |
| `request_size`             | The request's body size in bytes. | `kong.global.unmatched.request.size` |
| `response_size`            | The response's body size in bytes. | `kong.global.unmatched.response.size` |
| `latency`                  | The time interval between when the request started and when the response is received from the upstream server. | `kong.global.unmatched.latency` |
| `status_count`             | The status count. | `kong.global.unmatched.status.<status>.count` |
| `kong_latency`             | The internal {{site.base_gateway}} latency in milliseconds that it took to run all the plugins. | `kong.global.unmatched.kong_latency` |
```
I _think_ this might fix the table rendering

<!-- vale on -->
If you enable the `tag_style` configuration for the StatsD Plugin, the following metrics are sent instead:
<!-- vale off -->

> Metric                     | Description | Namespace
> ---                        | ---         | ---
> `request_count`            | The number of requests. | `kong.request.count`
> `request_size`             | The request's body size in bytes. | `kong.request.size`
> `response_size`            | The response's body size in bytes. | `kong.response.size`
> `latency`                  | The time interval in milliseconds between the request and response. | `kong.latency`
> `request_per_user`         | Tracks the request count per consumer. | `kong.request.count`
> `upstream_latency`         | Tracks the time in milliseconds it took for the final Service to process the request. | `kong.upstream_latency`
> `shdict_usage`             | The usage of shared dict, sent once every minute. | `kong.shdict.free_space` and `kong.shdict.capacity`
> `cache_datastore_hits_total`            | The total number of cache hits. ({{site.ee_product_name}} only) | `kong.cache_datastore_hits_total`
> `cache_datastore_misses_total`            | The total number of cache misses. ({{site.ee_product_name}} only) | `kong.cache_datastore_misses_total`

Reviewer: ```suggestion
| Metric                     | Description | Namespace |
| ---                        | ---         | --- |
| `request_count`            | The number of requests. | `kong.request.count` |
| `request_size`             | The request's body size in bytes. | `kong.request.size` |
| `response_size`            | The response's body size in bytes. | `kong.response.size` |
| `latency`                  | The time interval in milliseconds between the request and response. | `kong.latency` |
| `request_per_user`         | Tracks the request count per consumer. | `kong.request.count` |
| `upstream_latency`         | Tracks the time in milliseconds it took for the final Service to process the request. | `kong.upstream_latency` |
| `shdict_usage`             | The usage of shared dict, sent once every minute. | `kong.shdict.free_space` and `kong.shdict.capacity` |
| `cache_datastore_hits_total`            | The total number of cache hits. ({{site.base_gateway}} only) | `kong.cache_datastore_hits_total` |
| `cache_datastore_misses_total`            | The total number of cache misses. ({{site.base_gateway}} only) | `kong.cache_datastore_misses_total` |
```
I think the Enterprise labels can be replaced. I think this will fix the table rendering

<!-- vale on -->

The StatsD Plugin supports Librato, InfluxDB, DogStatsD, and SignalFX-style tags, which are used like Prometheus labels.

> * **Librato-style tags**: Must be appended to the metric name with a delimiting #, for example:

Reviewer: ```suggestion
* **Librato-style tags**: Must be appended to the metric name with a delimiting `#`, for example:
```

`metric.name#tagName=val,tag2Name=val2:0|c`
See the [Librato StatsD](https://github.com/librato/statsd-librato-backend#tags) documentation for more information.

* **InfluxDB-style tags**: Must be appended to the metric name with a delimiting comma, for example:
`metric.name,tagName=val,tag2Name=val2:0|c`
See the [InfluxDB StatsD](https://www.influxdata.com/blog/getting-started-with-sending-statsd-metrics-to-telegraf-influxdb/#introducing-influx-statsd) documentation for more information.

> * **DogStatsD-style tags**: Appended as a |# delimited section at the end of the metric, for example:

Reviewer: ```suggestion
* **DogStatsD-style tags**: Appended as a `|#` delimited section at the end of the metric, for example:
```

    datatype: String
    allowed: "One of the following options: `service_id`, `service_name`, `service_host`, `service_name_or_host`, `null`"
  - field: "`workspace_identifier`<br>*conditional*"
    description: Workspace detail.
    datatype: String
    allowed: "One of the following options: `workspace_id`, `workspace_name`, `null`"
{% endtable %}

<!-- vale on -->
### Metric behavior

- All metrics are logged by default.
- Metrics with a `stat_type` of `counter` or `gauge` **require** the `sample_rate` field.
- The `unique_users` metric only supports the `set` stat type.
- The following metrics only support the `counter` stat type:
  - `status_count`
  - `status_count_per_user`
  - `status_count_per_user_per_route`
  - `request_per_user`
- The `shdict_usage` metric only supports the `gauge` stat type.
- The following metrics **require** a `consumer_identifier`:
  - `status_count_per_user`
  - `request_per_user`
  - `unique_users`
  - `status_count_per_user_per_route`
- The `service_identifier` field is optional for all metrics. If not set, it defaults to `service_name_or_host`.
- The `status_count_per_workspace` metric requires a `workspace_identifier`.

> ## Kong Process Errors

Reviewer: ```suggestion
## {{site.base_gateway}} process errors
```

> This logging Plugin logs HTTP request and response data, and also supports stream data (TCP, TLS, and UDP).
> The Kong process error file is the Nginx error file. You can find it at the following path:
> `{prefix}/logs/error.log`

Reviewer: ```suggestion
This logging plugin logs HTTP request and response data, and also supports stream data (TCP, TLS, and UDP).
The {{site.base_gateway}} process error file is the Nginx error file. You can find it at the following path:
`{prefix}/logs/error.log`
```
